library/init/data/list/lemmas.lean

```diff
@@ -360,6 +382,171 @@ or.elim (mem_or_mem_of_mem_append this)
   (suppose a ∈ l₁, l₁subl this)
   (suppose a ∈ l₂, l₂subl this)
 
+/- foldr -/
+
+section foldr
+
+variable f : α → β → β
+
+@[simp]
+theorem foldr_nil (i : β) :
+  foldr f i nil = i
+:= rfl
+
+@[simp]
+theorem foldr_cons (i : β) (x : α) (xs : list α) :
+  foldr f i (x :: xs) = f x (foldr f i xs)
+:= rfl
+
+theorem foldr_append (f : α → β → β) :
+  ∀ (b : β) (l₁ l₂ : list α), foldr f b (l₁++l₂) = foldr f (foldr f b l₂) l₁
+| b nil         l₂ := rfl
+| b (cons a l₁) l₂ := congr_arg (f a) (foldr_append b l₁ l₂)
+
+end foldr
+
+section foldr_ac
```

Comment:
I could remove sum_foldr and sum_foldl. I put a bit more than we strictly need just so that it doesn't come out uneven. For instance, I tried to prove the same things about foldl and foldr. It allowed me to play around with the proofs and choose the one I prefer. I thought the unused theorems would still be worth keeping.